### ISSUE TITLE
rust-analyzer ignores type error

### ISSUE THREAD

**Issue by Spu7Nix** [rust-analyzer/rust-analyzer#6864]
I'm running rust-analyzer in vs code, and with this simple code it shows no error.

```rust
struct A;
struct B;

impl B {
    fn func(&self, m: A) {}
}

fn main() {
    let b = B {};
    // should show a type error
    // instead, shows nothing
    b.func("hi", "hello");
}

```

**Comment by Spu7Nix** [comment]
In fact, the (what I presume is a) bug appears in this even simpler code:
```rust
fn func(a: u8) {}
fn main() {
    func("a", "a");
}
```

**Comment by jonas-schievink** [comment]
This is currently working as intended. rust-analyzer does not display type mismatch diagnostics.

**Comment by Spu7Nix** [comment]
I might be misunderstanding, but it does report the error when there is only one agrument:
![image](https://user-images.githubusercontent.com/50080834/102080171-48656180-3e0e-11eb-9007-f79032021724.png)
```rust
fn func(a: u8) {}
fn main() {
    func("a");
}
```

**Comment by jonas-schievink** [comment]
Oh, those come from `cargo check`.

You're right, it really is missing for the example above, that's odd.

**Comment by jonas-schievink** [comment]
I don't immediately see anything wrong with the mapping we produce for this. This might be a VS Code bug.

<details>
<summary>Mapping</summary>

```
Diagnostic {
    message: "this function takes 1 argument but 2 arguments were supplied",
    code: Some(
        DiagnosticCode {
            code: "E0061",
            explanation: Some(
                "An invalid number of arguments was passed when calling a function.\n\nErroneous code example:\n\n```compile_fail,E0061\nfn f(u: i32) {}\n\nf(); // error!\n```\n\nThe number of arguments passed to a function must match the number of arguments\nspecified in the function signature.\n\nFor example, a function like:\n\n```\nfn f(a: u16, b: &str) {}\n```\n\nMust always be called with exactly two arguments, e.g., `f(2, \"test\")`.\n\nNote that Rust does not have a notion of optional function arguments or\nvariadic functions (except for its C-FFI).\n",
            ),
            __do_not_match_exhaustively: (),
        },
    ),
    level: Error,
    spans: [
        DiagnosticSpan {
            file_name: "crates/project_model/src/workspace.rs",
            byte_start: 11310,
            byte_end: 11313,
            line_start: 289,
            line_end: 289,
            column_start: 10,
            column_end: 13,
            is_primary: false,
            text: [
                DiagnosticSpanLine {
                    text: "    func(\"a\", \"a\");",
                    highlight_start: 10,
                    highlight_end: 13,
                    __do_not_match_exhaustively: (),
                },
            ],
            label: Some(
                "",
            ),
            suggested_replacement: None,
            suggestion_applicability: None,
            expansion: None,
            __do_not_match_exhaustively: (),
        },
        DiagnosticSpan {
            file_name: "crates/project_model/src/workspace.rs",
            byte_start: 11315,
            byte_end: 11318,
            line_start: 289,
            line_end: 289,
            column_start: 15,
            column_end: 18,
            is_primary: false,
            text: [
                DiagnosticSpanLine {
                    text: "    func(\"a\", \"a\");",
                    highlight_start: 15,
                    highlight_end: 18,
                    __do_not_match_exhaustively: (),
                },
            ],
            label: Some(
                "supplied 2 arguments",
            ),
            suggested_replacement: None,
            suggestion_applicability: None,
            expansion: None,
            __do_not_match_exhaustively: (),
        },
        DiagnosticSpan {
            file_name: "crates/project_model/src/workspace.rs",
            byte_start: 11271,
            byte_end: 11285,
            line_start: 287,
            line_end: 287,
            column_start: 1,
            column_end: 15,
            is_primary: false,
            text: [
                DiagnosticSpanLine {
                    text: "fn func(a: u8) {}",
                    highlight_start: 1,
                    highlight_end: 15,
                    __do_not_match_exhaustively: (),
                },
            ],
            label: Some(
                "defined here",
            ),
            suggested_replacement: None,
            suggestion_applicability: None,
            expansion: None,
            __do_not_match_exhaustively: (),
        },
        DiagnosticSpan {
            file_name: "crates/project_model/src/workspace.rs",
            byte_start: 11305,
            byte_end: 11309,
            line_start: 289,
            line_end: 289,
            column_start: 5,
            column_end: 9,
            is_primary: true,
            text: [
                DiagnosticSpanLine {
                    text: "    func(\"a\", \"a\");",
                    highlight_start: 5,
                    highlight_end: 9,
                    __do_not_match_exhaustively: (),
                },
            ],
            label: Some(
                "expected 1 argument",
            ),
            suggested_replacement: None,
            suggestion_applicability: None,
            expansion: None,
            __do_not_match_exhaustively: (),
        },
    ],
    children: [],
    rendered: Some(
        "error[E0061]: this function takes 1 argument but 2 arguments were supplied\n   --> crates/project_model/src/workspace.rs:289:5\n    |\n287 | fn func(a: u8) {}\n    | -------------- defined here\n288 | fn main() {\n289 |     func(\"a\", \"a\");\n    |     ^^^^ ---  --- supplied 2 arguments\n    |     |\n    |     expected 1 argument\n\n",
    ),
    __do_not_match_exhaustively: (),
} -> [
    MappedRustDiagnostic {
        url: Url {
            scheme: "file",
            host: None,
            port: None,
            path: "/home/jonas/dev/rust-analyzer/crates/project_model/src/workspace.rs",
            query: None,
            fragment: None,
        },
        diagnostic: Diagnostic {
            range: Range {
                start: Position {
                    line: 288,
                    character: 4,
                },
                end: Position {
                    line: 288,
                    character: 8,
                },
            },
            severity: Some(
                Error,
            ),
            code: Some(
                String(
                    "E0061",
                ),
            ),
            code_description: Some(
                CodeDescription {
                    href: Url {
                        scheme: "https",
                        host: Some(
                            Domain(
                                "doc.rust-lang.org",
                            ),
                        ),
                        port: None,
                        path: "/error-index.html",
                        query: None,
                        fragment: Some(
                            "E0061",
                        ),
                    },
                },
            ),
            source: Some(
                "rustc",
            ),
            message: "this function takes 1 argument but 2 arguments were supplied\nexpected 1 argument",
            related_information: Some(
                [
                    DiagnosticRelatedInformation {
                        location: Location {
                            uri: Url {
                                scheme: "file",
                                host: None,
                                port: None,
                                path: "/home/jonas/dev/rust-analyzer/crates/project_model/src/workspace.rs",
                                query: None,
                                fragment: None,
                            },
                            range: Range {
                                start: Position {
                                    line: 288,
                                    character: 9,
                                },
                                end: Position {
                                    line: 288,
                                    character: 12,
                                },
                            },
                        },
                        message: "",
                    },
                    DiagnosticRelatedInformation {
                        location: Location {
                            uri: Url {
                                scheme: "file",
                                host: None,
                                port: None,
                                path: "/home/jonas/dev/rust-analyzer/crates/project_model/src/workspace.rs",
                                query: None,
                                fragment: None,
                            },
                            range: Range {
                                start: Position {
                                    line: 288,
                                    character: 14,
                                },
                                end: Position {
                                    line: 288,
                                    character: 17,
                                },
                            },
                        },
                        message: "supplied 2 arguments",
                    },
                    DiagnosticRelatedInformation {
                        location: Location {
                            uri: Url {
                                scheme: "file",
                                host: None,
                                port: None,
                                path: "/home/jonas/dev/rust-analyzer/crates/project_model/src/workspace.rs",
                                query: None,
                                fragment: None,
                            },
                            range: Range {
                                start: Position {
                                    line: 286,
                                    character: 0,
                                },
                                end: Position {
                                    line: 286,
                                    character: 14,
                                },
                            },
                        },
                        message: "defined here",
                    },
                ],
            ),
            tags: None,
            data: None,
        },
        fixes: [],
    },
    MappedRustDiagnostic {
        url: Url {
            scheme: "file",
            host: None,
            port: None,
            path: "/home/jonas/dev/rust-analyzer/crates/project_model/src/workspace.rs",
            query: None,
            fragment: None,
        },
        diagnostic: Diagnostic {
            range: Range {
                start: Position {
                    line: 288,
                    character: 9,
                },
                end: Position {
                    line: 288,
                    character: 12,
                },
            },
            severity: Some(
                Hint,
            ),
            code: Some(
                String(
                    "E0061",
                ),
            ),
            code_description: Some(
                CodeDescription {
                    href: Url {
                        scheme: "https",
                        host: Some(
                            Domain(
                                "doc.rust-lang.org",
                            ),
                        ),
                        port: None,
                        path: "/error-index.html",
                        query: None,
                        fragment: Some(
                            "E0061",
                        ),
                    },
                },
            ),
            source: Some(
                "rustc",
            ),
            message: "",
            related_information: Some(
                [
                    DiagnosticRelatedInformation {
                        location: Location {
                            uri: Url {
                                scheme: "file",
                                host: None,
                                port: None,
                                path: "/home/jonas/dev/rust-analyzer/crates/project_model/src/workspace.rs",
                                query: None,
                                fragment: None,
                            },
                            range: Range {
                                start: Position {
                                    line: 288,
                                    character: 4,
                                },
                                end: Position {
                                    line: 288,
                                    character: 8,
                                },
                            },
                        },
                        message: "original diagnostic",
                    },
                ],
            ),
            tags: None,
            data: None,
        },
        fixes: [],
    },
    MappedRustDiagnostic {
        url: Url {
            scheme: "file",
            host: None,
            port: None,
            path: "/home/jonas/dev/rust-analyzer/crates/project_model/src/workspace.rs",
            query: None,
            fragment: None,
        },
        diagnostic: Diagnostic {
            range: Range {
                start: Position {
                    line: 288,
                    character: 14,
                },
                end: Position {
                    line: 288,
                    character: 17,
                },
            },
            severity: Some(
                Hint,
            ),
            code: Some(
                String(
                    "E0061",
                ),
            ),
            code_description: Some(
                CodeDescription {
                    href: Url {
                        scheme: "https",
                        host: Some(
                            Domain(
                                "doc.rust-lang.org",
                            ),
                        ),
                        port: None,
                        path: "/error-index.html",
                        query: None,
                        fragment: Some(
                            "E0061",
                        ),
                    },
                },
            ),
            source: Some(
                "rustc",
            ),
            message: "supplied 2 arguments",
            related_information: Some(
                [
                    DiagnosticRelatedInformation {
                        location: Location {
                            uri: Url {
                                scheme: "file",
                                host: None,
                                port: None,
                                path: "/home/jonas/dev/rust-analyzer/crates/project_model/src/workspace.rs",
                                query: None,
                                fragment: None,
                            },
                            range: Range {
                                start: Position {
                                    line: 288,
                                    character: 4,
                                },
                                end: Position {
                                    line: 288,
                                    character: 8,
                                },
                            },
                        },
                        message: "original diagnostic",
                    },
                ],
            ),
            tags: None,
            data: None,
        },
        fixes: [],
    },
    MappedRustDiagnostic {
        url: Url {
            scheme: "file",
            host: None,
            port: None,
            path: "/home/jonas/dev/rust-analyzer/crates/project_model/src/workspace.rs",
            query: None,
            fragment: None,
        },
        diagnostic: Diagnostic {
            range: Range {
                start: Position {
                    line: 286,
                    character: 0,
                },
                end: Position {
                    line: 286,
                    character: 14,
                },
            },
            severity: Some(
                Hint,
            ),
            code: Some(
                String(
                    "E0061",
                ),
            ),
            code_description: Some(
                CodeDescription {
                    href: Url {
                        scheme: "https",
                        host: Some(
                            Domain(
                                "doc.rust-lang.org",
                            ),
                        ),
                        port: None,
                        path: "/error-index.html",
                        query: None,
                        fragment: Some(
                            "E0061",
                        ),
                    },
                },
            ),
            source: Some(
                "rustc",
            ),
            message: "defined here",
            related_information: Some(
                [
                    DiagnosticRelatedInformation {
                        location: Location {
                            uri: Url {
                                scheme: "file",
                                host: None,
                                port: None,
                                path: "/home/jonas/dev/rust-analyzer/crates/project_model/src/workspace.rs",
                                query: None,
                                fragment: None,
                            },
                            range: Range {
                                start: Position {
                                    line: 288,
                                    character: 4,
                                },
                                end: Position {
                                    line: 288,
                                    character: 8,
                                },
                            },
                        },
                        message: "original diagnostic",
                    },
                ],
            ),
            tags: None,
            data: None,
        },
        fixes: [],
    },
]
```

</details>

**Comment by lnicola** [comment]
LSP Trace:

<details>
```
[Trace - 15:11:54] Received notification 'textDocument/publishDiagnostics'.
Params: {
    "uri": "file:///home/me/Projects/hello/src/main.rs",
    "diagnostics": [],
    "version": 4
}


[Trace - 15:11:54] Received notification 'textDocument/publishDiagnostics'.
Params: {
    "uri": "file:///home/me/Projects/hello/src/main.rs",
    "diagnostics": [
        {
            "range": {
                "start": {
                    "line": 3,
                    "character": 4
                },
                "end": {
                    "line": 3,
                    "character": 8
                }
            },
            "severity": 1,
            "code": "E0061",
            "codeDescription": {
                "href": "https://doc.rust-lang.org/error-index.html#E0061"
            },
            "source": "rustc",
            "message": "this function takes 1 argument but 2 arguments were supplied\nexpected 1 argument",
            "relatedInformation": [
                {
                    "location": {
                        "uri": "file:///home/me/Projects/hello/src/main.rs",
                        "range": {
                            "start": {
                                "line": 3,
                                "character": 9
                            },
                            "end": {
                                "line": 3,
                                "character": 12
                            }
                        }
                    },
                    "message": ""
                },
                {
                    "location": {
                        "uri": "file:///home/me/Projects/hello/src/main.rs",
                        "range": {
                            "start": {
                                "line": 3,
                                "character": 14
                            },
                            "end": {
                                "line": 3,
                                "character": 17
                            }
                        }
                    },
                    "message": "supplied 2 arguments"
                },
                {
                    "location": {
                        "uri": "file:///home/me/Projects/hello/src/main.rs",
                        "range": {
                            "start": {
                                "line": 0,
                                "character": 3
                            },
                            "end": {
                                "line": 0,
                                "character": 7
                            }
                        }
                    },
                    "message": "function defined here"
                }
            ]
        },
        {
            "range": {
                "start": {
                    "line": 3,
                    "character": 9
                },
                "end": {
                    "line": 3,
                    "character": 12
                }
            },
            "severity": 4,
            "code": "E0061",
            "codeDescription": {
                "href": "https://doc.rust-lang.org/error-index.html#E0061"
            },
            "source": "rustc",
            "message": "",
            "relatedInformation": [
                {
                    "location": {
                        "uri": "file:///home/me/Projects/hello/src/main.rs",
                        "range": {
                            "start": {
                                "line": 3,
                                "character": 4
                            },
                            "end": {
                                "line": 3,
                                "character": 8
                            }
                        }
                    },
                    "message": "original diagnostic"
                }
            ]
        },
        {
            "range": {
                "start": {
                    "line": 3,
                    "character": 14
                },
                "end": {
                    "line": 3,
                    "character": 17
                }
            },
            "severity": 4,
            "code": "E0061",
            "codeDescription": {
                "href": "https://doc.rust-lang.org/error-index.html#E0061"
            },
            "source": "rustc",
            "message": "supplied 2 arguments",
            "relatedInformation": [
                {
                    "location": {
                        "uri": "file:///home/me/Projects/hello/src/main.rs",
                        "range": {
                            "start": {
                                "line": 3,
                                "character": 4
                            },
                            "end": {
                                "line": 3,
                                "character": 8
                            }
                        }
                    },
                    "message": "original diagnostic"
                }
            ]
        },
        {
            "range": {
                "start": {
                    "line": 0,
                    "character": 3
                },
                "end": {
                    "line": 0,
                    "character": 7
                }
            },
            "severity": 4,
            "code": "E0061",
            "codeDescription": {
                "href": "https://doc.rust-lang.org/error-index.html#E0061"
            },
            "source": "rustc",
            "message": "function defined here",
            "relatedInformation": [
                {
                    "location": {
                        "uri": "file:///home/me/Projects/hello/src/main.rs",
                        "range": {
                            "start": {
                                "line": 3,
                                "character": 4
                            },
                            "end": {
                                "line": 3,
                                "character": 8
                            }
                        }
                    },
                    "message": "original diagnostic"
                }
            ]
        }
    ],
    "version": 4
}
```
</details>

**Comment by jonas-schievink** [comment]
Even weirder, this is showing up correctly, despite resulting in the same rustc error:

```rust
fn func(a: u8) {}
fn main() {
    func(0, 0);
}
```

**Comment by flodiebold** [comment]
It seems to look fine in Emacs, if I understand the problem correctly:
![2020-12-14-141731_1150x817_scrot](https://user-images.githubusercontent.com/906069/102086171-edd10300-3e17-11eb-8b73-174cfa8e58e2.png)

**Comment by Spu7Nix** [comment]
Could it be some problem in the front-end for the VSCode library?

**Comment by veber-alex** [comment]
Same issue happens with the "missing lifetime specifier" error.
The following code doesn't produce any error from cargo check:
```rust
fn either(a: &str, b: &str) -> &str {
    if true {
        a
    } else {
        b
    }
}
```

**Comment by JPMoresmau** [comment]
I have the same problem which seems to happen mainly when I have a closure:
```
struct Me();

impl Me {
    
    pub fn new() -> Self {
        Me().add(|n| n+1, "buzz")
    }

    pub fn add(mut self, matcher: i16) -> Self {
        self
    }

}
```
does not show any error in VSCode, while cargo check reports the error. 
The diagnostics message seems to contain several errors, I got something like:
```
[Trace - 12:34:01] Received notification 'textDocument/publishDiagnostics'.
Params: {
    "uri": "file:///c:/Users/jean-philippem/Exercism/rust/fizzy/src/lib.rs",
    "diagnostics": [
        {
            "range": {
                "start": {
                    "line": 82,
                    "character": 13
                },
                "end": {
                    "line": 82,
                    "character": 16
                }
            },
            "severity": 1,
            "code": "E0061",
            "codeDescription": {
                "href": "https://doc.rust-lang.org/error-index.html#E0061"
            },
            "source": "rustc",
            "message": "this function takes 1 argument but 2 arguments were supplied\nexpected 1 argument",
            "relatedInformation": [
                {
                    "location": {
                        "uri": "file:///c:/Users/jean-philippem/Exercism/rust/fizzy/src/lib.rs",
                        "range": {
                            "start": {
                                "line": 82,
                                "character": 17
                            },
                            "end": {
                                "line": 82,
                                "character": 24
                            }
                        }
                    },
                    "message": ""
                },
                {
                    "location": {
                        "uri": "file:///c:/Users/jean-philippem/Exercism/rust/fizzy/src/lib.rs",
                        "range": {
                            "start": {
                                "line": 82,
                                "character": 25
                            },
                            "end": {
                                "line": 82,
                                "character": 31
                            }
                        }
                    },
                    "message": "supplied 2 arguments"
                },
                {
                    "location": {
                        "uri": "file:///c:/Users/jean-philippem/Exercism/rust/fizzy/src/lib.rs",
                        "range": {
                            "start": {
                                "line": 85,
                                "character": 4
                            },
                            "end": {
                                "line": 85,
                                "character": 46
                            }
                        }
                    },
                    "message": "defined here"
                }
            ]
        },
        {
            "range": {
                "start": {
                    "line": 82,
                    "character": 17
                },
                "end": {
                    "line": 82,
                    "character": 24
                }
            },
            "severity": 4,
            "code": "E0061",
            "codeDescription": {
                "href": "https://doc.rust-lang.org/error-index.html#E0061"
            },
            "source": "rustc",
            "message": "",
            "relatedInformation": [
                {
                    "location": {
                        "uri": "file:///c:/Users/jean-philippem/Exercism/rust/fizzy/src/lib.rs",
                        "range": {
                            "start": {
                                "line": 82,
                                "character": 13
                            },
                            "end": {
                                "line": 82,
                                "character": 16
                            }
                        }
                    },
                    "message": "original diagnostic"
                }
            ]
        },
        {
            "range": {
                "start": {
                    "line": 82,
                    "character": 25
                },
                "end": {
                    "line": 82,
                    "character": 31
                }
            },
            "severity": 4,
            "code": "E0061",
            "codeDescription": {
                "href": "https://doc.rust-lang.org/error-index.html#E0061"
            },
            "source": "rustc",
            "message": "supplied 2 arguments",
            "relatedInformation": [
                {
                    "location": {
                        "uri": "file:///c:/Users/jean-philippem/Exercism/rust/fizzy/src/lib.rs",
                        "range": {
                            "start": {
                                "line": 82,
                                "character": 13
                            },
                            "end": {
                                "line": 82,
                                "character": 16
                            }
                        }
                    },
                    "message": "original diagnostic"
                }
            ]
        },
        {
            "range": {
                "start": {
                    "line": 85,
                    "character": 4
                },
                "end": {
                    "line": 85,
                    "character": 46
                }
            },
            "severity": 4,
            "code": "E0061",
            "codeDescription": {
                "href": "https://doc.rust-lang.org/error-index.html#E0061"
            },
            "source": "rustc",
            "message": "defined here",
            "relatedInformation": [
                {
                    "location": {
                        "uri": "file:///c:/Users/jean-philippem/Exercism/rust/fizzy/src/lib.rs",
                        "range": {
                            "start": {
                                "line": 82,
                                "character": 13
                            },
                            "end": {
                                "line": 82,
                                "character": 16
                            }
                        }
                    },
                    "message": "original diagnostic"
                }
            ]
        }
    ],
    "version": 135
}
```
But it doesn't show up in VSCode. Not sure how I can debug this further.

**Comment by lnicola** [comment]
Probably duplicate of https://github.com/rust-analyzer/rust-analyzer/issues/6892.
